### ISSUE TITLE
Joystick with only axes and no buttons not recognised

### ISSUE THREAD

**Issue by broekema** [pygame/pygame#2932]
**Environment:**
Arch linux (2021-12-19)
Python 3.10.1
sdl2 2.0.18-1
pygame 2.1.0 (SDL 2.0.16, Python 3.10.1)

Hardware: self-built controller running QMK on a Arduino pro micro. I am using 4 potentiometers that are configured as axes, and 12 "buttons" that act as ordinary keys. 

**Current behavior:**

When the joystick says it has 4 axes but no buttons, it is not detected by pygame.
When the joystick is reconfigured to pretend it has at least one button, it is detected by pygame.
Note that the SDL test program *does* detect the joystick without buttons, suggesting that this is an edge case in pygame itself, not in the underlying SDL library.

**Expected behavior:**

Regardless of the number of buttons, pygame should detect the controller.

**Steps to reproduce:**

Might be slightly inconvenient without a custom "joystick" but I'd be happy to provide a working QMK .hex file.

**Comment by MyreMylar** [comment]
As far as I can tell there isn't anything in the pygame code joystick code that tests for the presence of a specific number of buttons.

Since you are familiar with making an SDL program yourself you can probably look it over yourself and see:

https://github.com/pygame/pygame/blob/main/src_c/joystick.c

It might also be useful to see a simple pygame program you are using that demonstrates the issue and a simple SDL program that shows the same.

I assume you are actually using the same version of SDL for the pygame and pure SDL test?

Usually hardware joystick issues turn out to be because the particular joystick in question has not yet been added to SDL's big list of joystick PIDs & VIDs here:

https://github.com/libsdl-org/SDL/blob/main/src/joystick/controller_type.h

**Comment by broekema** [comment]
Everything is using the system SDL library, so I'm assuming the same versions are used. Obviously the custom device is not in the big list of joystick PID & VIDs, neither is it in the gamecontrollerdb.txt file.

I understand there is no test for a number of buttons in the actual python code, and I didn't expect that to be the problem. What I do expect is that there is some weird edge-case that fails to properly initialise the joystick object when no buttons are defined.

Simple test program to illustrate the issue:
```
import pygame

pygame.init();
pygame.joystick.init();

for i in range(0, pygame.joystick.get_count()):
    joystick = pygame.joystick.Joystick(i)
    if joystick.get_guid() == "03000000efbe00006969000011010000":
       print("Found controller named ", joystick.get_name(), " with ", joystick.get_numaxes(), " axes and ", joystick.get_numbuttons(), " buttons.");

pygame.quit()
```

Running this code with no buttons configured:
```
% python test.py                                                                            
pygame 2.1.0 (SDL 2.0.16, Python 3.10.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
% 
```
As indicated, the device is not detected. There is a `/dev/` entry that is properly populated, and both the native joystick test program and SDL2 joystick test program work as expected.

Native joystick test program:
```
% jstest /dev/input/js0
Driver version is 2.1.0.
Joystick (QMK handwired/misterdeck) has 4 axes (X, Y, Z, Rx)
and 4 buttons (Trigger, ThumbBtn, ThumbBtn2, TopBtn).
Testing ... (interrupt to exit)
Axes:  0:-28377  1:-29053  2:-29053  3:-28039 Buttons:  0:off  1:off  2:off  3:off ^C
% 
```

SDL test program with the same configuration:
```
% sdl2-jstest --list      
error: failed to read gamecontrollerdb.txt: Invalid RWops
Found 1 joystick(s)

Joystick Name:     'QMK handwired/misterdeck'
Joystick GUID:     03000000efbe00006969000011010000
Joystick Number:    0
Number of Axes:     4
Number of Buttons:  0
Number of Hats:     0
Number of Balls:    0
GameControllerConfig:
  missing (see 'gamecontrollerdb.txt' or SDL_GAMECONTROLLERCONFIG)
% 
```

Reconfigured the device with 4 joystick buttons:
```
% python test.py    
pygame 2.1.0 (SDL 2.0.16, Python 3.10.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
Found controller named  QMK handwired/misterdeck  with  4  axes and  4  buttons.
% 
```

Corresponding sdl2 test program:
```
% sdl2-jstest --list                            
error: failed to read gamecontrollerdb.txt: Invalid RWops
Found 1 joystick(s)

Joystick Name:     'QMK handwired/misterdeck'
Joystick GUID:     03000000efbe00006969000011010000
Joystick Number:    0
Number of Axes:     4
Number of Buttons:  4
Number of Hats:     0
Number of Balls:    0
GameControllerConfig:
  missing (see 'gamecontrollerdb.txt' or SDL_GAMECONTROLLERCONFIG)
```

**Comment by MyreMylar** [comment]
What is the C code for the `sdl2-jstest` program you are using? Perhaps we could see if there is any difference from the SDL functions pygame is calling in `joystick.c`

**Comment by MyreMylar** [comment]
Also; does `joystick.get_guid()` return something when you run this program:

```
import pygame

pygame.init();
pygame.joystick.init();

print("num joysticks found: ", pygame.joystick.get_count())

for i in range(0, pygame.joystick.get_count()):
    joystick = pygame.joystick.Joystick(i)
    print (joystick.get_guid())

pygame.quit()
```

On your no button's hardware setup?

**Comment by broekema** [comment]
> Also; does joystick.get_guid() return something when you run this program:
```
% python test2.py 
pygame 2.1.0 (SDL 2.0.16, Python 3.10.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
num joysticks found:  0
```

> What is the C code for the sdl2-jstest program you are using?

The sdl2-jstest program is in gitlab here: [https://gitlab.com/sdl-jstest/sdl-jstest/-/blob/master/sdl-jstest.c](https://gitlab.com/sdl-jstest/sdl-jstest/-/blob/master/sdl-jstest.c)

**Comment by ankith26** [comment]
Hmm, seems like `joystick.get_count()` returns a `0` in pygame even when the joystick is connected. This function uses `SDL_NumJoysticks` under the hood, does that have the same output in C code?

**Comment by broekema** [comment]
You are right, my C example is using a different SDL version (2.0.18-3). I had not realised pygame packages its own SDL library. Digging a bit deeper, this does seem to appear to cause the issue. I copied the packaged SDL2 library from pygame (and renamed it a bit for convenience).

Using the SDL2 library packaged with pygame the joystick is not detected:
```
chris@wintendo ~/projects/StreamSliders % sdl2-jstest --list
error: failed to read gamecontrollerdb.txt: Invalid RWops
No joysticks were found
chris@wintendo ~/projects/StreamSliders % ldd /usr/bin/sdl2-jstest 
	linux-vdso.so.1 (0x00007ffc645f8000)
	libSDL2-2.0.so.0 => /home/chris/projects/StreamSliders/libSDL2-2.0.so.0 (0x00007f69c0aff000)
	libncursesw.so.6 => /usr/lib/libncursesw.so.6 (0x00007f69c0a41000)
	libc.so.6 => /usr/lib/libc.so.6 (0x00007f69c0875000)
	libm.so.6 => /usr/lib/libm.so.6 (0x00007f69c0731000)
	libdl.so.2 => /usr/lib/libdl.so.2 (0x00007f69c072a000)
	libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00007f69c0709000)
	librt.so.1 => /usr/lib/librt.so.1 (0x00007f69c06fc000)
	/lib64/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007f69c0ca9000)
chris@wintendo ~/projects/StreamSliders % 
```
When using the system SDL2 library, the joystick is found as can be expected.
```
chris@wintendo ~ % sdl2-jstest --list
Found 1 joystick(s)

Joystick Name:     'QMK MisteRdeck'
Joystick GUID:     03000000efbe00006969000011010000
Joystick Number:    0
Number of Axes:     4
Number of Buttons:  0
Number of Hats:     0
Number of Balls:    0
GameControllerConfig:
  missing (see 'gamecontrollerdb.txt' or SDL_GAMECONTROLLERCONFIG)

chris@wintendo ~ % ldd /usr/bin/sdl2-jstest
	linux-vdso.so.1 (0x00007ffd9b8ca000)
	libSDL2-2.0.so.0 => /usr/lib/libSDL2-2.0.so.0 (0x00007fb6cf1a9000)
	libncursesw.so.6 => /usr/lib/libncursesw.so.6 (0x00007fb6cf136000)
	libc.so.6 => /usr/lib/libc.so.6 (0x00007fb6cef6a000)
	libm.so.6 => /usr/lib/libm.so.6 (0x00007fb6cee26000)
	libdl.so.2 => /usr/lib/libdl.so.2 (0x00007fb6cee1f000)
	libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00007fb6cedfe000)
	/lib64/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007fb6cf3a2000)
chris@wintendo ~ % 
```
So this is an upstream issue that is already fixed. Provided the SDL2 library packaged with pygame is updated in the next release, I think we can close this issue.

**Comment by MyreMylar** [comment]
The latest version of pygame (2.1.2) released last week should now have SDL 2.0.18 so hopefully it already fixed if you do something like:

`pip3 install pygame -U`

**Comment by broekema** [comment]
I just updated pygame to version 2.1.2, but there is no difference in the SDL version that I can see (md5sum is identical, and the libraries are named identically). I'm not sure SDL 2.0.18 got shipped.

**Comment by MyreMylar** [comment]
Well it made it in on windows. You should be able to see the version of SDL when you run any basic pygame program.

Looks like this:

```
pygame 2.1.2 (SDL 2.0.18, Python 3.9.0)
```

I'm not sure what the `SDL 2.0.18-1`, `SDL 2.0.18-3` things are though. They don't appear to be official versions of SDL - Is Arch Linux making their own patches to SDL? Does pygame not dynamically grab the latest SDL runtime on linux? Or is that a thing that end users need to do. See:

https://www.reddit.com/r/linux_gaming/comments/1upn39/sdl2_adds_dynamic_api_magic_to_allow_updating_it/

**Comment by MyreMylar** [comment]
Looking into it, the answer is yes ArchLinux roll their own patches to SDL with the numbering system. I don't think pygame can do anything about this.

**Comment by broekema** [comment]
I agree that we should not worry about the Arch linux packages. 

However, it seems something weird is going on with pygame 2.1.2 as distributed. 
```
chris@wintendo ~ % pip3 cache purge     
ERROR: No matching packages
1 chris@wintendo ~ % pip3 install pygame
Defaulting to user installation because normal site-packages is not writeable
Collecting pygame
  Downloading pygame-2.1.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (21.9 MB)
     |████████████████████████████████| 21.9 MB 9.0 MB/s            
Installing collected packages: pygame
Successfully installed pygame-2.1.2
chris@wintendo ~ % python3              
Python 3.10.1 (main, Dec 18 2021, 23:53:45) [GCC 11.1.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pygame
pygame 2.1.2 (SDL 2.0.16, Python 3.10.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
>>> 
```
That said, I'll close this issue since it is an upstream bug that seems to be fixed in SDL 2.0.18.

**Comment by MyreMylar** [comment]
Yeah, poking into the .whl file:

`libSDL2-2-302c83d0.0.so.0.16.0`

it looks like the many linux wheel builds didn't get SDL 2.0.18 like the windows and mac versions did. The build script on Github is up to date so must have been a small snafu somewhere in the building process. Hopefully in the next release.
